providers/azure.go

```diff
@@ -230,44 +276,48 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 		err = otherMailsErr
 	}
 
+	if err != nil || email == "" {
+		email, err = json.Get("userPrincipalName").String()
+		if err != nil {
+			logger.Errorf("unable to find userPrincipalName: %s", err)
+			return "", err
+		}
+	}
```

Comment:
Long term, this feels a bit weird. I would have thought as an admin you would want to be deterministic about where the email comes from rather than having it try 3 different possible fields

Reply:
Agreed - but this looks to be the existing functionality?

I'm more curious after seeing the ADFS PR #1021 where that contributor wanted to use the `upn` claim as the `session.User` how people want to consume that field in general?

Is an alternative to the traditional `session.User = claim.Subject` something that is in demand? 🤔

Reply:
AAD has two versions and each has its own set of optional claims. Even id_token is optional. :) So it's highly configurable. In general, for v1 format, `upn` makes sense. I don't know the history of adopting `userPrincipalName`, though. But I think we should at least keep it as compatible as it is?

Reply:
`session.User = claim.Subject` would be a bad idea on Azure AD as it's a pairwise identifier - it is unique to a particular application ID.
https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#payload-claims

Reply:
Agreed - for your PR, lets keep this existing logic backwards compatible and we can revisit it later in a breaking release if we want 👍